api/oneclick.py

```diff
@@ -185,12 +341,167 @@ def get_patron_internal_id(self, patron_email=None, patron_cardno=None):
 
         resp_dict = response.json()
         message = resp_dict.get('message', None)
-        error_response = self.validate_response(response, message, action=action)
+        try:
+            self.validate_response(response, message, action=action)
+        except PatronNotFoundException, e:
+            # this should not be fatal at this point
+            return None
 
-        internal_patron_id = resp_dict['patronId']
+        internal_patron_id = resp_dict.get('patronId', None)
         return internal_patron_id
 
 
+    def get_patron_checkouts(self, patron_id):
+        """
+        Gets the books and audio the patron currently has checked out.
+        Obtains fulfillment info for each item -- the way to fulfill a book 
+        is to get this list of possibilities first, and then call individual 
+        fulfillment endpoints on the individual items.
+
+        :param patron_id OneClick internal id for the patron.
+        """
+        url = "%s/libraries/%s/patrons/%s/checkouts/" % (self.base_url, str(self.library_id), patron_id)
+        action="patron_checkouts"
+        loans = []
+
+        resp_obj = []
+        message = None
+        try:
+            response = self.request(url=url)
+
+            if response.text:
+                resp_obj = response.json()
+                # if we succeeded, then we got back a list of checkouts
+                # if we failed, then we got back a dictionary with an error message
+                if isinstance(resp_obj, dict):
+                    message = resp_obj.get('message', None)
+        except Exception, e:
+            self.log.error("Patron checkouts failed: %r", e, exc_info=e)
+            raise RemoteInitiatedServerError(e.message, action)
+
+        self.validate_response(response=response, message=message, action=action)
+
+        # by now we can assume response is either empty or a list
+        for item in resp_obj:
+            # go through patron's checkouts and generate LoanInfo objects, 
+            # with FulfillmentInfo objects included
+            media_type = item.get('mediaType', 'eBook')
+            isbn = item.get('isbn', None)
+            can_renew = item.get('canRenew', None)
+            title = item.get('title', None)
+            authors = item.get('authors', None)
+            # refers to checkout expiration date, not the downloadUrl's
+            expires = item.get('expiration', None)
+            if expires:
+                expires = datetime.datetime.strptime(expires, self.EXPIRATION_DATE_FORMAT).date()
+
+            identifier, made_new = Identifier.for_foreign_id(self._db, 
+                    foreign_identifier_type=Identifier.ONECLICK_ID, 
+                    foreign_id=isbn, autocreate=False)
+
+            # Note: if OneClick knows about a patron's checked-out item that wasn't
+            # checked out through us, we ignore it
+            if not identifier:
+                continue
```

Comment:
This isn't consistent with how we treat books checked out via the Overdrive app--you can see those books but not necessarily fulfill them. I guess it doesn't matter because we're creating a totally new One Click account for the patron which is used solely by SimplyE, so it won't come up.